File: src/main/cpp/commands/PassBallCommand.cpp
#include "commands/PassBallCommand.h"

#include <cmath>

#include <frc/DriverStation.h>
#include <frc/smartdashboard/SmartDashboard.h>

#include "Constants.h"
#include "shooting/ShotTable.h"

using namespace units::literals;

PassBallCommand::PassBallCommand(
    subsystems::CommandSwerveDrivetrain* drive,
    subsystems::ShooterSubsystem* shooter,
    subsystems::FeederSubsystem* feeder,
    units::degree_t shooterFacingOffset)
    : drive_(drive),
      shooter_(shooter),
      feeder_(feeder),
      shooter_facing_offset_(shooterFacingOffset) {
  AddRequirements({drive, shooter, feeder});
}

void PassBallCommand::Initialize() {
  timer_started_ = false;
  feeding_ = false;
  shot_timer_.Stop();
  shot_timer_.Reset();
  feeder_->Stop();

  if (shooter_->GetPitchHomeState() ==
          subsystems::ShooterSubsystem::PitchHomeState::kUnhomed ||
      shooter_->GetPitchHomeState() ==
          subsystems::ShooterSubsystem::PitchHomeState::kFault) {
    shooter_->BeginPitchHoming();
  }

  drive_->SOMangleDiff = 180.0;
  facing_request_.WithHeadingPID(9, 0, 0.1)
      .WithDeadband(max_speed_ * 0.05)
      .WithRotationalDeadband(0.1_rad_per_s)
      .WithMaxAbsRotationalRate(3.14_rad_per_s)
      .WithDriveRequestType(swerve::DriveRequestType::Velocity)
      .WithSteerRequestType(swerve::SteerRequestType::Position);
}

void PassBallCommand::Execute() {
  const auto alliance = frc::DriverStation::GetAlliance();
  if (!alliance.has_value()) {
    feeder_->Stop();
    shooter_->SetIdle();
    return;
  }

  const bool is_red = alliance.value() == frc::DriverStation::Alliance::kRed;
  const frc::Pose2d pose = drive_->GetState().Pose;
  frc::Translation2d target =
      pose.Y() > 4_m ? blue_left_target_ : blue_right_target_;
  if (is_red) {
    target = frc::Translation2d{FieldConstants::kFieldLength - target.X(),
                                target.Y()};
  }

  const auto target_direction =
      frc::Rotation2d{units::math::atan2(target.Y() - pose.Y(),
                                         target.X() - pose.X())} -
      frc::Rotation2d{shooter_facing_offset_};
  const double angle_error =
      std::abs((target_direction - pose.Rotation()).Degrees().value());
  drive_->SOMangleDiff = angle_error;

  if (angle_error < 2.0) {
    drive_->SetControl(brake_request_);
  } else {
    drive_->SetControl(facing_request_.WithVelocityX(0_mps)
                           .WithVelocityY(0_mps)
                           .WithTargetDirection(target_direction));
  }

  if (!shooter_->IsPitchHomed()) {
    feeder_->SetBackwardFeederDuty(0.0);
    feeder_->SetUpwardDuty(0.0);
    return;
  }

  const auto distance = pose.Translation().Distance(target);
  const auto setpoint = shooting::ShotTable::Pass(distance);
  shooter_->ApplyShotSetpoint(setpoint);
  feeder_->SetUpwardFeederVelocity(setpoint.upperFeeder.value());

  if (!timer_started_) {
    shot_timer_.Restart();
    timer_started_ = true;
  }

  const bool clear_of_hub = pose.Y() <= 3.5_m || pose.Y() >= 4.5_m;
  const bool ready = clear_of_hub && angle_error <= 3.0 &&
                     shooter_->IsFlywheelReady(0.7_tps) &&
                     shooter_->IsPitchReady(0.75_deg) &&
                     std::abs(feeder_->GetUpwardFeederVelocity() -
                              setpoint.upperFeeder.value()) <= 2.0;

  if (shooting::ShotTable::FeedAllowed(
          shooter_->IsPitchHomed(), ready,
          shot_timer_.HasElapsed(1.5_s))) {
    feeding_ = true;
  }
  feeder_->SetBackwardFeederDuty(feeding_ ? 1.0 : 0.0);

  frc::SmartDashboard::PutNumber("Shooting/PassDistanceM", distance.value());
  frc::SmartDashboard::PutBoolean("Shooting/PassReady", ready);
  frc::SmartDashboard::PutBoolean("Shooting/PassForcedFeed",
                                  feeding_ && !ready);
}

void PassBallCommand::End(bool interrupted) {
  shot_timer_.Stop();
  feeder_->Stop();
  shooter_->SetIdle();
  drive_->SOMangleDiff = 180.0;

  const auto speeds = drive_->GetState().Speeds;
  drive_->SetControl(drive_->m_safeCoastRequest.WithVelocityX(speeds.vx)
                         .WithVelocityY(speeds.vy)
                         .WithRotationalRate(speeds.omega));
}

bool PassBallCommand::IsFinished() { return false; }

File: src/main/java/frc/robot/commands/PassBallCommand.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.commands;

import static edu.wpi.first.units.Units.*;

import java.util.Optional;

import com.ctre.phoenix6.swerve.SwerveModule.DriveRequestType;
import com.ctre.phoenix6.swerve.SwerveModule.SteerRequestType;
import com.ctre.phoenix6.swerve.SwerveRequest;

import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Translation2d;
import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.DriverStation.Alliance;
import edu.wpi.first.wpilibj.Timer;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.Command;

import frc.robot.Constants;
import frc.robot.generated.TunerConstants;
import frc.robot.shooting.ShotSetpoint;
import frc.robot.shooting.ShotTable;
import frc.robot.subsystems.CommandSwerveDrivetrain;
import frc.robot.subsystems.FeederSubsystem;
import frc.robot.subsystems.ShooterSubsystem;

public class PassBallCommand extends Command {
    private final CommandSwerveDrivetrain drive;
    private final ShooterSubsystem shooter;
    private final FeederSubsystem feeder;
    private final double shooterFacingOffsetDeg;
    private boolean timerStarted = false;
    private boolean feeding = false;
    private final Timer shotTimer = new Timer();

    private final double maxSpeedMetersPerSecond =
            TunerConstants.kSpeedAt12Volts.in(MetersPerSecond);
    private final SwerveRequest.FieldCentricFacingAngle facingRequest =
            new SwerveRequest.FieldCentricFacingAngle();

    private final Translation2d blueLeftTarget = new Translation2d(3.0, 5.5);
    private final Translation2d blueRightTarget = new Translation2d(3.0, 2.5);

    public PassBallCommand(
            CommandSwerveDrivetrain drive,
            ShooterSubsystem shooter,
            FeederSubsystem feeder) {
        this(drive, shooter, feeder, 180.0);
    }

    public PassBallCommand(
            CommandSwerveDrivetrain drive,
            ShooterSubsystem shooter,
            FeederSubsystem feeder,
            double shooterFacingOffsetDeg) {
        this.drive = drive;
        this.shooter = shooter;
        this.feeder = feeder;
        this.shooterFacingOffsetDeg = shooterFacingOffsetDeg;
        addRequirements(drive, shooter, feeder);
    }

    @Override
    public void initialize() {
        timerStarted = false;
        feeding = false;
        shotTimer.stop();
        shotTimer.reset();
        feeder.stop();

        if (shooter.getPitchHomeState() == ShooterSubsystem.PitchHomeState.UNHOMED
                || shooter.getPitchHomeState() == ShooterSubsystem.PitchHomeState.FAULT) {
            shooter.beginPitchHoming();
        }

        drive.setSomAngleDiff(180.0);
        facingRequest
                .withHeadingPID(9.0, 0.0, 0.1)
                .withDeadband(maxSpeedMetersPerSecond * 0.05)
                .withRotationalDeadband(0.1)
                .withMaxAbsRotationalRate(3.14)
                .withDriveRequestType(DriveRequestType.Velocity)
                .withSteerRequestType(SteerRequestType.Position);
    }

    @Override
    public void execute() {
        Optional<Alliance> alliance = DriverStation.getAlliance();
        if (alliance.isEmpty()) {
            feeder.stop();
            shooter.setIdle();
            return;
        }

        boolean isRed = alliance.get() == Alliance.Red;
        Pose2d pose = drive.getState().Pose;
        Translation2d target = pose.getY() > 4.0 ? blueLeftTarget : blueRightTarget;
        if (isRed) {
            target = new Translation2d(
                    Constants.FieldConstants.fieldLengthMeters - target.getX(),
                    target.getY());
        }

        Rotation2d targetDirection = Rotation2d
                .fromRadians(Math.atan2(target.getY() - pose.getY(), target.getX() - pose.getX()))
                .minus(Rotation2d.fromDegrees(shooterFacingOffsetDeg));
        double angleErrorDeg = Math.abs(targetDirection.minus(pose.getRotation()).getDegrees());
        drive.setSomAngleDiff(angleErrorDeg);

        if (angleErrorDeg < 2.0) {
            drive.setBrakeRequest();
        } else {
            drive.setControl(
                    facingRequest
                            .withVelocityX(0.0)
                            .withVelocityY(0.0)
                            .withTargetDirection(targetDirection));
        }

        if (!shooter.isPitchHomed()) {
            feeder.setBackwardFeederDuty(0.0);
            feeder.setUpwardDuty(0.0);
            return;
        }

        double distance = pose.getTranslation().getDistance(target);
        ShotSetpoint setpoint = ShotTable.pass(distance);
        shooter.applyShotSetpoint(setpoint);
        feeder.setUpwardFeederVelocity(setpoint.feederRps());

        if (!timerStarted) {
            shotTimer.restart();
            timerStarted = true;
        }

        boolean clearOfHub = pose.getY() <= 3.5 || pose.getY() >= 4.5;
        boolean ready = clearOfHub
                && angleErrorDeg <= 3.0
                && shooter.isFlywheelReady(0.7)
                && shooter.isPitchReady(0.75)
                && Math.abs(feeder.getUpwardFeederVelocity() - setpoint.feederRps()) <= 2.0;

        if (ShotTable.feedAllowed(shooter.isPitchHomed(), ready, shotTimer.hasElapsed(1.5))) {
            feeding = true;
        }
        feeder.setBackwardFeederDuty(feeding ? 1.0 : 0.0);

        SmartDashboard.putNumber("Shooting/PassDistanceM", distance);
        SmartDashboard.putBoolean("Shooting/PassReady", ready);
        SmartDashboard.putBoolean("Shooting/PassForcedFeed", feeding && !ready);
    }

    @Override
    public void end(boolean interrupted) {
        shotTimer.stop();
        feeder.stop();
        shooter.setIdle();
        drive.setSomAngleDiff(180.0);
        drive.setSafeCoastWithCurrentSpeeds();
    }

    @Override
    public boolean isFinished() {
        return false;
    }
}
